File: objc/Tests/TestPerAUFile.m
/*
 * TestPerAUFile — v1.0 file-level per-AU encryption round-trip.
 *
 * Mirrors python/tests/test_per_au_file.py. Builds a plaintext
 * fixture, calls TTIOPerAUFile.encryptFilePath:... (with and
 * without header encryption), reopens, verifies the file carries
 * the new feature flags and compound datasets, then decrypts and
 * compares plaintext signal values bit-for-bit.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#import <Foundation/Foundation.h>
#import "Testing.h"
#import <unistd.h>
#import <string.h>

#import "Protection/TTIOPerAUFile.h"
#import "Protection/TTIOPerAUEncryption.h"
#import "Dataset/TTIOSpectralDataset.h"
#import "Dataset/TTIOWrittenRun.h"
#import "Providers/TTIOProviderRegistry.h"
#import "Providers/TTIOStorageProtocols.h"
#import "HDF5/TTIOFeatureFlags.h"
#import "HDF5/TTIOHDF5File.h"
#import "HDF5/TTIOHDF5Group.h"
#import "ValueClasses/TTIOEnums.h"
#import "Run/TTIOAcquisitionRun.h"
#import "Spectra/TTIOSpectrum.h"
#import "Core/TTIOSignalArray.h"


static NSString *tmpPath(NSString *n) {
    return [NSString stringWithFormat:@"/tmp/ttio_peraufile_%d_%@",
            (int)getpid(), n];
}
static void rmFile(NSString *p) { [[NSFileManager defaultManager] removeItemAtPath:p error:NULL]; }
static NSData *key42(void) {
    uint8_t b[32]; memset(b, 0x42, 32);
    return [NSData dataWithBytes:b length:32];
}

static NSData *f64arr(const double *v, NSUInteger n) {
    NSMutableData *d = [NSMutableData dataWithCapacity:n * 8];
    for (NSUInteger i = 0; i < n; i++) [d appendBytes:&v[i] length:8];
    return d;
}
static NSData *u64arr(const uint64_t *v, NSUInteger n) {
    NSMutableData *d = [NSMutableData dataWithCapacity:n * 8];
    for (NSUInteger i = 0; i < n; i++) [d appendBytes:&v[i] length:8];
    return d;
}
static NSData *u32arr(const uint32_t *v, NSUInteger n) {
    NSMutableData *d = [NSMutableData dataWithCapacity:n * 4];
    for (NSUInteger i = 0; i < n; i++) [d appendBytes:&v[i] length:4];
    return d;
}
static NSData *i32arr(const int32_t *v, NSUInteger n) {
    NSMutableData *d = [NSMutableData dataWithCapacity:n * 4];
    for (NSUInteger i = 0; i < n; i++) [d appendBytes:&v[i] length:4];
    return d;
}

static BOOL buildPlaintextFixture(NSString *path, NSError **error)
{
    NSUInteger n = 5, p = 4, total = n * p;
    double mz[20], intensity[20];
    for (NSUInteger i = 0; i < total; i++) {
        mz[i] = 100.0 + (double)i;
        intensity[i] = (double)(i + 1) * 10.0;
    }
    uint64_t offsets[5] = {0, 4, 8, 12, 16};
    uint32_t lengths[5] = {4, 4, 4, 4, 4};
    double rts[5] = {1.5, 3.0, 4.5, 6.0, 7.5};
    int32_t msLevels[5] = {1, 2, 1, 2, 1};
    int32_t pols[5] = {1, 1, 1, 1, 1};
    double pmzs[5] = {0.0, 501.0, 0.0, 503.0, 0.0};
    int32_t pcs[5] = {0, 2, 0, 2, 0};
    double bpis[5];
    for (NSUInteger i = 0; i < n; i++) {
        double best = 0.0;
        for (NSUInteger k = 0; k < p; k++) {
            double v = intensity[i * p + k];
            if (v > best) best = v;
        }
        bpis[i] = best;
    }
    TTIOWrittenRun *run =
        [[TTIOWrittenRun alloc]
            initWithSpectrumClassName:@"TTIOMassSpectrum"
                      acquisitionMode:(int64_t)TTIOAcquisitionModeMS1DDA
                          channelData:@{@"mz": f64arr(mz, total),
                                        @"intensity": f64arr(intensity, total)}
                              offsets:u64arr(offsets, n)
                              lengths:u32arr(lengths, n)
                       retentionTimes:f64arr(rts, n)
                             msLevels:i32arr(msLevels, n)
                           polarities:i32arr(pols, n)
                         precursorMzs:f64arr(pmzs, n)
                     precursorCharges:i32arr(pcs, n)
                  basePeakIntensities:f64arr(bpis, n)];
    return [TTIOSpectralDataset writeMinimalToPath:path
                                              title:@"per-AU file round-trip fixture"
                                 isaInvestigationId:@"ISA-OBJC-PERAU"
                                             msRuns:@{@"run_0001": run}
                                    identifications:nil
                                    quantifications:nil
                                  provenanceRecords:nil
                                              error:error];
}


// Helper: reopen the HDF5 file to inspect feature flags + children
// without going through TTIOSpectralDataset (which may refuse
// post-encryption).
static NSArray<NSString *> *readFeaturesFromFile(NSString *path)
{
    TTIOHDF5File *f = [TTIOHDF5File openReadOnlyAtPath:path error:NULL];
    TTIOHDF5Group *root = f.rootGroup;
    NSArray *feats = [TTIOFeatureFlags featuresForRoot:root];
    return feats ?: @[];
}


void testPerAUFile(void)
{
    // ── 1. Encrypt (channels only) writes segments + flag ───────
    {
        NSString *path = tmpPath(@"src1.tio");
        rmFile(path);
        NSError *err = nil;
        PASS(buildPlaintextFixture(path, &err), "fixture built");

        BOOL ok = [TTIOPerAUFile encryptFilePath:path
                                              key:key42()
                                 encryptHeaders:NO
                                    providerName:nil
                                           error:&err];
        PASS(ok, "encryptFilePath (channels only) succeeds");

        NSArray *features = readFeaturesFromFile(path);
        PASS([features containsObject:@"opt_per_au_encryption"],
             "opt_per_au_encryption flag set");
        PASS(![features containsObject:@"opt_encrypted_au_headers"],
             "opt_encrypted_au_headers flag NOT set (channels-only mode)");
        rmFile(path);
    }

    // ── 2. Encrypt + decrypt recovers plaintext ─────────────────
    {
        NSString *path = tmpPath(@"src2.tio");
        rmFile(path);
        NSError *err = nil;
        buildPlaintextFixture(path, &err);

        // Capture original plaintext bytes before encryption mutates
        // the file.
        TTIOSpectralDataset *src = [TTIOSpectralDataset readFromFilePath:path error:&err];
        TTIOAcquisitionRun *run = src.msRuns[@"run_0001"];
        NSMutableData *origMz = [NSMutableData data];
        NSMutableData *origInt = [NSMutableData data];
        for (NSUInteger i = 0; i < [run count]; i++) {
            TTIOSpectrum *sp = [run objectAtIndex:i];
            [origMz appendData:sp.signalArrays[@"mz"].buffer];
            [origInt appendData:sp.signalArrays[@"intensity"].buffer];
        }
        // Release the read handle before we reopen for in-place
        // encryption; TTIOSpectralDataset holds the HDF5 file open
        // otherwise.
        [src closeFile];
        src = nil;

        BOOL ok = [TTIOPerAUFile encryptFilePath:path
                                              key:key42()
                                 encryptHeaders:NO
                                    providerName:nil
                                           error:&err];
        PASS(ok, "encryptFilePath succeeds on fresh fixture");

        err = nil;
        NSDictionary *decrypted =
            [TTIOPerAUFile decryptFilePath:path
                                         key:key42()
                                providerName:nil
                                       error:&err];
        PASS(decrypted != nil, "decryptFilePath succeeds");
        NSDictionary *runOut = decrypted[@"run_0001"];
        PASS([runOut[@"mz"] isEqualToData:origMz],
             "mz bytes round-trip bit-for-bit");
        PASS([runOut[@"intensity"] isEqualToData:origInt],
             "intensity bytes round-trip bit-for-bit");
        rmFile(path);
    }

    // ── 3. Wrong key fails decrypt ──────────────────────────────
    {
        NSString *path = tmpPath(@"src3.tio");
        rmFile(path);
        NSError *err = nil;
        buildPlaintextFixture(path, &err);
        [TTIOPerAUFile encryptFilePath:path
                                     key:key42()
                        encryptHeaders:NO
                           providerName:nil
                                  error:&err];
        uint8_t badBytes[32]; memset(badBytes, 0, 32);
        NSData *bad = [NSData dataWithBytes:badBytes length:32];
        err = nil;
        NSDictionary *result =
            [TTIOPerAUFile decryptFilePath:path
                                         key:bad
                                providerName:nil
                                       error:&err];
        PASS(result == nil && err != nil, "wrong key rejected (AES-GCM auth)");
        rmFile(path);
    }

    // ── 4. encryptHeaders=YES writes au_header_segments + flag ──
    {
        NSString *path = tmpPath(@"src4.tio");
        rmFile(path);
        NSError *err = nil;
        buildPlaintextFixture(path, &err);

        BOOL ok = [TTIOPerAUFile encryptFilePath:path
                                              key:key42()
                                 encryptHeaders:YES
                                    providerName:nil
                                           error:&err];
        PASS(ok, "encryptFilePath (with headers) succeeds");

        NSArray *features = readFeaturesFromFile(path);
        PASS([features containsObject:@"opt_per_au_encryption"],
             "opt_per_au_encryption set (headers mode)");
        PASS([features containsObject:@"opt_encrypted_au_headers"],
             "opt_encrypted_au_headers set (headers mode)");

        NSDictionary *decrypted =
            [TTIOPerAUFile decryptFilePath:path
                                         key:key42()
                                providerName:nil
                                       error:&err];
        PASS(decrypted != nil, "decryptFilePath w/ headers succeeds");
        NSArray *headers = decrypted[@"run_0001"][@"__au_headers__"];
        PASS(headers.count == 5, "__au_headers__ has 5 rows");
        if (headers.count == 5) {
            TTIOAUHeaderPlaintext *h1 = headers[1];
            PASS(h1.msLevel == 2 && h1.precursorMz == 501.0,
                 "row 1 semantic fields recovered (ms_level=2, pmz=501.0)");
            PASS(fabs(h1.retentionTime - 3.0) < 1e-12,
                 "row 1 retention_time recovered (3.0)");
        }
        rmFile(path);
    }
}

File: python/tests/test_per_au_file.py
"""v1.0 per-AU encryption file-level round-trip tests."""
from __future__ import annotations

from pathlib import Path

import numpy as np
import pytest

pytest.importorskip("cryptography")
pytest.importorskip("h5py")

import h5py

from ttio.enums import AcquisitionMode, Polarity
from ttio.encryption_per_au import decrypt_per_au_file, encrypt_per_au_file
from ttio.feature_flags import (
    OPT_ENCRYPTED_AU_HEADERS,
    OPT_PER_AU_ENCRYPTION,
)
from ttio.spectral_dataset import SpectralDataset, WrittenRun


KEY = b"\xAB" * 32  # deterministic test key


def _make_plaintext(path: Path, n_spectra: int = 5,
                     points_per_spectrum: int = 4) -> Path:
    total = n_spectra * points_per_spectrum
    mz = np.arange(total, dtype="<f8") + 100.0
    intensity = (np.arange(total, dtype="<f8") + 1.0) * 10.0
    offsets = np.arange(0, total, points_per_spectrum, dtype="<u8")
    lengths = np.full(n_spectra, points_per_spectrum, dtype="<u4")
    rts = np.arange(n_spectra, dtype="<f8") * 1.5
    ms_levels = np.array(
        [1 if i % 2 == 0 else 2 for i in range(n_spectra)], dtype="<i4"
    )
    polarities = np.full(n_spectra, int(Polarity.POSITIVE), dtype="<i4")
    precursor_mzs = np.array(
        [0.0 if ms_levels[i] == 1 else 500.0 + i for i in range(n_spectra)],
        dtype="<f8",
    )
    precursor_charges = np.array(
        [0 if ms_levels[i] == 1 else 2 for i in range(n_spectra)], dtype="<i4"
    )
    base_peak = np.array(
        [float(intensity[i * points_per_spectrum:(i + 1) * points_per_spectrum].max())
         for i in range(n_spectra)],
        dtype="<f8",
    )
    run = WrittenRun(
        spectrum_class="TTIOMassSpectrum",
        acquisition_mode=int(AcquisitionMode.MS1_DDA),
        channel_data={"mz": mz, "intensity": intensity},
        offsets=offsets,
        lengths=lengths,
        retention_times=rts,
        ms_levels=ms_levels,
        polarities=polarities,
        precursor_mzs=precursor_mzs,
        precursor_charges=precursor_charges,
        base_peak_intensities=base_peak,
    )
    SpectralDataset.write_minimal(
        path,
        title="per-AU encryption fixture",
        isa_investigation_id="ISA-PERAU",
        runs={"run_0001": run},
    )
    return path


# ---------------------------------------------------------- channel-only


class TestPerAuEncryptionChannels:

    def test_encrypt_writes_segments_and_sets_flag(self, tmp_path):
        path = _make_plaintext(tmp_path / "src.tio")
        encrypt_per_au_file(str(path), KEY)

        with h5py.File(str(path), "r") as f:
            sig = f["study/ms_runs/run_0001/signal_channels"]
            assert "mz_segments" in sig
            assert "intensity_segments" in sig
            assert "mz_values" not in sig
            assert "intensity_values" not in sig
            # Algorithm attribute attached.
            algo = sig.attrs["mz_algorithm"]
            if isinstance(algo, bytes):
                algo = algo.decode()
            assert algo == "aes-256-gcm"
            features_json = f.attrs.get("ttio_features", b"[]")
            if isinstance(features_json, bytes):
                features_json = features_json.decode()
            assert OPT_PER_AU_ENCRYPTION in features_json
            assert OPT_ENCRYPTED_AU_HEADERS not in features_json

    def test_decrypt_recovers_plaintext_values(self, tmp_path):
        path = _make_plaintext(tmp_path / "src.tio")
        # Capture original plaintext before encrypt mutates the file.
        with SpectralDataset.open(path) as ds:
            run = ds.all_runs["run_0001"]
            originals = {
                c: np.concatenate([
                    np.asarray(run[i].signal_array(c).data)
                    for i in range(len(run))
                ])
                for c in run.channel_names
            }
        encrypt_per_au_file(str(path), KEY)
        recovered = decrypt_per_au_file(str(path), KEY)["run_0001"]
        for c in originals:
            np.testing.assert_array_equal(originals[c], recovered[c])

    def test_wrong_key_fails_decrypt(self, tmp_path):
        path = _make_plaintext(tmp_path / "src.tio")
        encrypt_per_au_file(str(path), KEY)
        bad_key = b"\x00" * 32
        with pytest.raises(Exception):
            decrypt_per_au_file(str(path), bad_key)


# ---------------------------------------------------------- with headers


class TestPerAuEncryptionWithHeaders:

    def test_encrypt_headers_writes_au_header_segments(self, tmp_path):
        path = _make_plaintext(tmp_path / "src.tio")
        encrypt_per_au_file(str(path), KEY, encrypt_headers=True)
        with h5py.File(str(path), "r") as f:
            idx = f["study/ms_runs/run_0001/spectrum_index"]
            assert "au_header_segments" in idx
            for plain_name in ("retention_times", "ms_levels", "polarities",
                                "precursor_mzs", "precursor_charges",
                                "base_peak_intensities"):
                assert plain_name not in idx
            # lengths stays plaintext (structural). v1.10 #10: offsets
            # is no longer written on disk; it's computed from cumsum.
            assert "offsets" not in idx
            assert "lengths" in idx
            features_json = f.attrs.get("ttio_features", b"[]")
            if isinstance(features_json, bytes):
                features_json = features_json.decode()
            assert OPT_ENCRYPTED_AU_HEADERS in features_json
            assert OPT_PER_AU_ENCRYPTION in features_json

    def test_decrypt_with_headers_recovers_fields(self, tmp_path):
        path = _make_plaintext(tmp_path / "src.tio", n_spectra=3)
        # Capture originals.
        with h5py.File(str(path), "r") as f:
            idx = f["study/ms_runs/run_0001/spectrum_index"]
            orig_rts = idx["retention_times"][...]
            orig_ms = idx["ms_levels"][...]
            orig_pmz = idx["precursor_mzs"][...]

        encrypt_per_au_file(str(path), KEY, encrypt_headers=True)
        recovered = decrypt_per_au_file(str(path), KEY)["run_0001"]
        headers = recovered["__au_headers__"]
        assert len(headers) == 3
        for i in range(3):
            assert headers[i]["retention_time"] == pytest.approx(float(orig_rts[i]))
            assert headers[i]["ms_level"] == int(orig_ms[i])
            assert headers[i]["precursor_mz"] == pytest.approx(float(orig_pmz[i]))

    def test_row_swap_rejected_end_to_end(self, tmp_path):
        """Swap two rows of the on-disk channel segments. Decrypt must
        reject (AAD binds ciphertext to au_sequence)."""
        path = _make_plaintext(tmp_path / "src.tio", n_spectra=3)
        encrypt_per_au_file(str(path), KEY)
        with h5py.File(str(path), "r+") as f:
            segs_ds = f["study/ms_runs/run_0001/signal_channels/intensity_segments"]
            arr = segs_ds[...]
            arr[[0, 1]] = arr[[1, 0]]
            segs_ds[...] = arr
        with pytest.raises(Exception):
            decrypt_per_au_file(str(path), KEY)
